besu/src/main/java/org/hyperledger/besu/controller/IbftBesuControllerBuilder.java

```diff
@@ -215,16 +226,39 @@ protected void validateContext(final ProtocolContext<IbftContext> context) {
   @Override
   protected IbftContext createConsensusContext(
       final Blockchain blockchain, final WorldStateArchive worldStateArchive) {
-    final IbftConfigOptions ibftConfig =
-        genesisConfig.getConfigOptions(genesisConfigOverrides).getIbft2ConfigOptions();
+    final GenesisConfigOptions configOptions =
+        genesisConfig.getConfigOptions(genesisConfigOverrides);
+    final IbftConfigOptions ibftConfig = configOptions.getIbft2ConfigOptions();
     final EpochManager epochManager = new EpochManager(ibftConfig.getEpochLength());
+
+    final Map<Long, List<Address>> ibftValidatorForkMap =
+        convertIbftForks(configOptions.getCustomForks().getIbftForks());
+
     return new IbftContext(
-        new VoteTallyCache(
+        new ForkingVoteTallyCache(
             blockchain,
             new VoteTallyUpdater(epochManager, new IbftBlockInterface()),
             epochManager,
-            new IbftBlockInterface()),
+            new IbftBlockInterface(),
+            ibftValidatorForkMap),
         new VoteProposer(),
         blockInterface);
   }
+
+  private Map<Long, List<Address>> convertIbftForks(final List<IbftFork> ibftForks) {
```

Comment:
I wonder if instead of using the long to list mapping, we introduce an object that would be used by `ForkingVoteTallyCache`. Like `IbftValidatorsOverride` (probably could use a better name).

Using this mapping kinda implies an understanding that the block height is being used as the marker points for the forks. The map works though, not a big deal.

The logic in `ForkingVoteTallyCache ` would look like:

```
if (validatorsOverride.existsForBlock(actualBlockNumber)) {
      return new VoteTally(validatorsOverride.getForBlock(actualBlockNumber));
}
```

Reply:
The data can be encapsulated pretty cheaply - don't love having 2 functions 1 to check, and 1 to do the work ... but that can be worked around.